### ISSUE TITLE
Multiple copies of the same element with the same name cause problems

### ISSUE THREAD

**Issue by mfsilva22** [PrairieLearn/PrairieLearn#1278]
When users don't want to show the correct-answers in the answer-panel, they can move the corresponding input element (for example `pl-number-input`) inside `pl-question-panel`. This works, but also prevents the students to see the history of their submissions. Unfortunately, adding a copy of the element inside `pl-submission-panel` breaks the code. 

My suggestions:
1) Add text to the documentation to clarify this. It was not intuitive to me that I could not have the element inside submission and question panels. I could not find anything in the documentation, so I had to do look into the element code to understand that this was a limitation of the implementation, and not something I was doing incorrect as a user. I think a sentence or two in the documentation may help some other users in the future.

2) The first suggestion solves more the immediate issue in my understanding. For the future, it would be nice to be able to have an element showing in both question and submission panels

**Comment by mfsilva22** [comment]
Yes, your example above was exactly what I observed (thank you for providing a much better explanation of the issue I was reporting :-)). I also checked with other elements, such as the variations of matrix input. I told Matt that I thought I found a bug, and he asked me to open this report. I agree with you that it looks more like a bug, than just a implementation decision. But if the "bug" (or whatever we want to name) does not get fixed, I think we should address this limitation in the documentation.

**Comment by nicknytko** [comment]
The issue here looks like that because there are two of the same "element", it will parse the input twice (and probably grade, test, etc. as well).  A workaround for this is to use `<pl-hide-in-panel>` so that there is only one instance of the input element, but for the future it may be worth looking into combining duplicates like this?

**Comment by mwest1066** [comment]
Thanks for investigating! And good point, we basically don't handle multiple elements accessing the same data. Is fixing this just a matter of modifying each element so they check whether they've already parsed/graded/etc and return if so? For example, `parse()` could check whether there is already a `submitted_answer` entry with the specified name and bail out if so.

Or if we aren't going to handle the repeated-element case intelligently then we should probably detect and throw an error.

I agree that this isn't really related to #2065, so we might want to handle it separately from that PR.

**Comment by nicknytko** [comment]
That might be tricky to do since most elements take input from `submitted_answers` and then overwrite it with the parsed version.  Just a thought, but perhaps we can go a level up into the JS and scan for unique elements, then force `parse`, `grade`, etc. to only be called once per unique answer/elem.

**Comment by mwest1066** [comment]
Good point on `submitted_answers`, I was forgetting that we populate it by default from `raw_submitted_answers`, so we can't use that to detect processing. I like the idea of handling this at one level up. The trick is figuring out what we mean by a "unique" element. Most input elements have a `"answers-name"` attribute or something like this that makes them unique, but this is just a convention and I'm not sure all elements stick to it.

Perhaps we could make a new convention for elements to figure out whether they are unique or not? For example, we could set `data["processed_ids"] = {}` every time that we are about to call all the element functions. Then any element that cares (mainly input elements) could do something like this:
```python
my_answers_name = pl.get_attribute("answers_name")
if my_answers_name in data["processed_ids"]:
    # we are a duplicate element
    return

# do the actual processing
```

Or maybe in the `info.json` file for the element we could add a `id_attribute_name`, which would often be `answers-name`. Then the upper-level JS code could use this to track element uniqueness and only call one copy of each element (for prepare/grade/etc). I guess I'm back to your idea here 😄

**Comment by eliotwrobson** [comment]
I think this issue can be closed, since it's essentially a duplicate of #5881 in terms of code changes. For the specific situation described in the original issue, this can be solved by setting show-answer-panel to False in info.json

**Comment by eliotwrobson** [comment]
Going to go ahead and close this issue, as in https://github.com/PrairieLearn/PrairieLearn/pull/5893, we added documentation to prevent weird error messages for multiple copies of the same element (now checking all `answers-name` attributes are distinct), and updated the documentation accordingly.
